contracts/staking/GoodFundManager.sol

```diff
@@ -189,76 +237,127 @@ contract GoodFundManager is DAOContract {
+        }else{
+            require(interest >= 4 * gasPriceIncDAI, "Collected interests should be at least 4 times bigger than gas cost since last call of this function sooner than 2 months");
+        }
         // Mints gd while the interest amount is equal to the transferred amount
-        uint256 gdUBI = GoodReserveCDai(reserveAddress).mintUBI(
+        (uint256 gdUBI) = GoodReserveCDai(nameService.addresses(nameService.RESERVE())).mintUBI(
             iToken,
-            interest
+            interest // interest
         );
-        //_staking.updateGlobalGDYieldPerToken(gdInterest, interest);
         // Transfers the minted tokens to the given staking contract
         IGoodDollar token = IGoodDollar(address(avatar.nativeToken()));
-        //if(gdInterest > 0)
-        //    require(token.transfer(address(_staking), gdInterest),"interest transfer failed");
+        
         if(gdUBI > 0)
             //transfer ubi to avatar on sidechain via bridge
             require(token.transferAndCall(
                 bridgeContract,
                 gdUBI,
                 abi.encodePacked(ubiRecipient)
             ),"ubi bridge transfer failed");
+        uint256 gasCostToMintReward = 200000; // Gas cost to mint GD reward to keeper hardcoded so should be changed according to calculations
+        uint256 totalUsedGas = (initialGas - gasleft() + gasCostToMintReward) * 110 / 100; // We will return as reward 1.1x of used gas in GD 
+        uint256 gdAmountToMint= getGasPriceInGD(totalUsedGas);
+        GoodReserveCDai(nameService.addresses(nameService.RESERVE())).mintRewardFromRR(nameService.getAddress("CDAI"),msg.sender,gdAmountToMint);
+        lastCollectedInterest = block.timestamp;
         emit FundsTransferred(
             msg.sender,
-            address(_staking),
-            reserveAddress,
+            nameService.addresses(nameService.RESERVE()),
             interest,
-            gdUBI
+            gdUBI,
+            gdAmountToMint
         );
     }
+
+    /**
+     * @dev  Function that get addresses of staking contracts which interests 
+     * can be collected with the gas amount that provided as parameter
+     * @param _maxGasAmount The maximum amount of the gas that keeper willing to spend collect interests
+     * @return addresses of the staking contracts to the collect interests
+     */
+    function calcSortedContracts(uint256 _maxGasAmount) public view returns(address[] memory){
+        uint activeContractsLength = activeContracts.length;
+        address[] memory addresses = new address[](activeContractsLength); 
+        uint256[] memory balances = new uint256[](activeContractsLength);
+        uint256 tempInterest;
+        uint totalInterest;
+        uint i;
+        require(activeContractsLength > 0 , "There should be at least one active staking contract");
+        for (i = 0; i < activeContractsLength; i++){
+            (tempInterest, ,) = StakingContract(activeContracts[i]).currentUBIInterest();
```

Comment:
check that contract is actually still active, ie not passed block end, reward>0, not blacklisted

Reply:
Yes but if some interest stayed in balance even tho blockend passed shouldn't we collect it?

Reply:
ok you are right